File: system-test/dotnet/Driver.Adapter/External/Clock/Client/ClockStubClient.cs
using Driver.Adapter.External.Clock.Client.Dtos;

using Driver.Adapter.External.Clock.Client.Dtos.Error;

using Driver.Adapter.Shared.Client.Http;

using Common;

using Driver.Adapter.Shared.Client.WireMock;



namespace Driver.Adapter.External.Clock.Client;



public class ClockStubClient : IDisposable

{

    private const string HealthEndpoint = "/health";

    private const string TimeEndpoint = "/api/time";

    private const string ClockTimeEndpoint = "/clock/api/time";



    private readonly JsonHttpClient<ExtClockErrorResponse> _httpClient;

    private readonly JsonWireMockClient _wireMockClient;

    private bool _disposed;



    public ClockStubClient(string baseUrl)

    {

        _httpClient = new JsonHttpClient<ExtClockErrorResponse>(baseUrl);

        _wireMockClient = new JsonWireMockClient(baseUrl);

    }



    public void Dispose()

    {

        Dispose(true);

        GC.SuppressFinalize(this);

    }



    protected virtual void Dispose(bool disposing)

    {

        if (_disposed) return;

        if (disposing)

        {

            _httpClient.Dispose();

            _wireMockClient.Dispose();

        }

        _disposed = true;

    }



    public Task<Result<VoidValue, ExtClockErrorResponse>> CheckHealthAsync()

        => _httpClient.GetAsync(HealthEndpoint);



    public Task<Result<ExtGetTimeResponse, ExtClockErrorResponse>> GetTimeAsync()

        => _httpClient.GetAsync<ExtGetTimeResponse>(TimeEndpoint);



    public Task<Result<VoidValue, ExtClockErrorResponse>> ConfigureGetTimeAsync(ExtGetTimeResponse response)

        => _wireMockClient.StubGetAsync(ClockTimeEndpoint, HttpStatus.Ok, response)

            .MapErrorAsync(ExtClockErrorResponse.From);

}

File: system-test/java/src/main/java/com/optivem/shop/dsl/driver/adapter/external/clock/client/ClockStubClient.java
package com.optivem.shop.dsl.driver.adapter.external.clock.client;

import com.optivem.shop.dsl.driver.adapter.shared.client.http.HttpStatus;
import com.optivem.shop.dsl.driver.adapter.shared.client.http.JsonHttpClient;
import com.optivem.shop.dsl.common.Closer;
import com.optivem.shop.dsl.common.Result;
import com.optivem.shop.dsl.driver.adapter.shared.client.wiremock.JsonWireMockClient;
import com.optivem.shop.dsl.driver.adapter.external.clock.client.dtos.ExtGetTimeResponse;
import com.optivem.shop.dsl.driver.adapter.external.clock.client.dtos.error.ExtClockErrorResponse;

public class ClockStubClient implements AutoCloseable {
    private static final String HEALTH_ENDPOINT = "/health";
    private static final String TIME_ENDPOINT = "/api/time";
    private static final String CLOCK_TIME_ENDPOINT = "/clock/api/time";

    private final JsonHttpClient<ExtClockErrorResponse> httpClient;

    private final JsonWireMockClient wireMockClient;

    public ClockStubClient(String baseUrl) {
        this.httpClient = new JsonHttpClient<>(baseUrl, ExtClockErrorResponse.class);
        this.wireMockClient = new JsonWireMockClient(baseUrl);
    }

    @Override
    public void close() {
        Closer.close(httpClient);
    }

    public Result<Void, ExtClockErrorResponse> checkHealth() {
        return httpClient.get(HEALTH_ENDPOINT);
    }

    public Result<ExtGetTimeResponse, ExtClockErrorResponse> getTime() {
        return httpClient.get(TIME_ENDPOINT, ExtGetTimeResponse.class);
    }

    public Result<Void, ExtClockErrorResponse> configureGetTime(ExtGetTimeResponse response) {
        return wireMockClient.stubGet(CLOCK_TIME_ENDPOINT, HttpStatus.OK, response)
                .mapError(ExtClockErrorResponse::new);
    }

}
